ufl/algorithms/apply_derivatives.py

```diff
@@ -545,13 +545,16 @@ def argument(self, o):
     def reference_value(self, o):
         # grad(o) == grad(rv(f)) -> K_ji*rgrad(rv(f))_rj
         f = o.ufl_operands[0]
-        if f.ufl_element().mapping() == "physical":
+        el = f.ufl_function_space().ufl_element()
+        if el.mapping() == "physical":
             # TODO: Do we need to be more careful for immersed things?
             return ReferenceGrad(o)
 
         if not f._ufl_is_terminal_:
             error("ReferenceValue can only wrap a terminal")
-        domain = f.ufl_domain()
+
+        domains = f.ufl_function_space().ufl_domains()
+        domain = None if len(domains) == 0 else domains[0]
```

Comment:
I think it should be the case that a function space always has a domain? Should we instead write this as:

```
domain, = f.ufl_function_space().ufl_domains()
```

To assert that there is only one.